File: cpp/22_custom_data_structures.cpp
// ============================================================
// Topic:  Custom Data Structures  (FOCUS: linked list, binary tree, graph)
// Summary: Pointer-based list, tree traversals, BFS/DFS — built from scratch.
// Run:     g++ -std=c++17 cpp/22_custom_data_structures.cpp -o cpp/22_custom_data_structures && cpp/22_custom_data_structures
//
// KEY DIFFERENCES:
//   Lists:  Python class Node | TS class Node | Java static Node | unique_ptr<Node>
//   Trees:  object refs / pointers | Box<Option<TreeNode>> (Rust) | unique_ptr<TreeNode>
//   Graphs: adjacency list dict/list | vector<vector<int>> + BFS queue / DFS stack
// ============================================================

#include <iostream>
#include <memory>
#include <queue>
#include <vector>

// ----- 1. Singly linked list — unique_ptr owns each node -----
struct ListNode {
    int value;
    std::unique_ptr<ListNode> next;

    explicit ListNode(int v) : value(v), next(nullptr) {}
};

class SinglyLinkedList {
    std::unique_ptr<ListNode> head;

public:
    void push_back(int value) {
        auto* tail = head.get();
        if (!tail) {
            head = std::make_unique<ListNode>(value);
            return;
        }
        while (tail->next) {
            tail = tail->next.get();
        }
        tail->next = std::make_unique<ListNode>(value);
    }

    void reverse() {
        std::unique_ptr<ListNode> prev = nullptr;
        ListNode* current = head.release();
        while (current) {
            ListNode* next = current->next.release();
            current->next = std::move(prev);
            prev.reset(current);
            current = next;
        }
        head = std::move(prev);
    }

    void print(const char* label) const {
        std::cout << label;
        for (ListNode* node = head.get(); node; node = node->next.get()) {
            std::cout << (node == head.get() ? ": " : " ") << node->value;
        }
        std::cout << std::endl;
    }
};

// ----- 2. Binary tree — unique_ptr for left/right children -----
struct TreeNode {
    int value;
    std::unique_ptr<TreeNode> left;
    std::unique_ptr<TreeNode> right;

    explicit TreeNode(int v) : value(v), left(nullptr), right(nullptr) {}
};

class BinaryTree {
    std::unique_ptr<TreeNode> root;

    void inorder(const TreeNode* node, std::vector<int>& out) const {
        if (!node) return;
        inorder(node->left.get(), out);
        out.push_back(node->value);
        inorder(node->right.get(), out);
    }

    void postorder(const TreeNode* node, std::vector<int>& out) const {
        if (!node) return;
        postorder(node->left.get(), out);
        postorder(node->right.get(), out);
        out.push_back(node->value);
    }

    void preorder(const TreeNode* node, std::vector<int>& out) const {
        if (!node) return;
        out.push_back(node->value);
        preorder(node->left.get(), out);
        preorder(node->right.get(), out);
    }

    static void printTraversal(const char* label, const std::vector<int>& values) {
        std::cout << label;
        for (size_t i = 0; i < values.size(); ++i) {
            std::cout << (i == 0 ? ": " : " ") << values[i];
        }
        std::cout << std::endl;
    }

public:
    BinaryTree() {
        // Tree:     1
        //          / \
        //         2   3
        //        / \
        //       4   5
        root = std::make_unique<TreeNode>(1);
        root->left = std::make_unique<TreeNode>(2);
        root->right = std::make_unique<TreeNode>(3);
        root->left->left = std::make_unique<TreeNode>(4);
        root->left->right = std::make_unique<TreeNode>(5);
    }

    void printTraversals() const {
        std::vector<int> values;
        values.clear();
        inorder(root.get(), values);
        printTraversal("inorder", values);
        values.clear();
        postorder(root.get(), values);
        printTraversal("postorder", values);
        values.clear();
        preorder(root.get(), values);
        printTraversal("preorder", values);
    }
};

// ----- 3. Graph — adjacency list + BFS queue / DFS recursion -----
class Graph {
    std::vector<std::vector<int>> adjacency;

    void dfsVisit(int vertex, std::vector<bool>& visited, std::vector<int>& order) const {
        visited[vertex] = true;
        order.push_back(vertex);
        for (int neighbor : adjacency[vertex]) {
            if (!visited[neighbor]) {
                dfsVisit(neighbor, visited, order);
            }
        }
    }

    static void printOrder(const char* label, const std::vector<int>& order) {
        std::cout << label;
        for (size_t i = 0; i < order.size(); ++i) {
            std::cout << (i == 0 ? ": " : " ") << order[i];
        }
        std::cout << std::endl;
    }

public:
    explicit Graph(int vertexCount) : adjacency(vertexCount) {}

    void addEdge(int from, int to) { adjacency[from].push_back(to); }

    void bfs(int start) const {
        std::vector<bool> visited(adjacency.size(), false);
        std::queue<int> frontier;
        std::vector<int> order;

        visited[start] = true;
        frontier.push(start);

        while (!frontier.empty()) {
            int vertex = frontier.front();
            frontier.pop();
            order.push_back(vertex);
            for (int neighbor : adjacency[vertex]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    frontier.push(neighbor);
                }
            }
        }
        printOrder("bfs", order);
    }

    void dfs(int start) const {
        std::vector<bool> visited(adjacency.size(), false);
        std::vector<int> order;
        dfsVisit(start, visited, order);
        printOrder("dfs", order);
    }
};

int main() {
    // ----- 1. Linked list — push, print, reverse -----
    SinglyLinkedList list;
    for (int value : {1, 2, 3, 4, 5}) {
        list.push_back(value);
    }
    list.print("list");
    // Output: list: 1 2 3 4 5
    list.reverse();
    list.print("reversed");
    // Output: reversed: 5 4 3 2 1

    // ----- 2. Binary tree — inorder, postorder, preorder -----
    BinaryTree tree;
    tree.printTraversals();
    // Output: inorder: 4 2 5 1 3
    // Output: postorder: 4 5 2 3 1
    // Output: preorder: 1 2 4 5 3

    // ----- 3. Graph — BFS and DFS from vertex 0 -----
    Graph graph(5);
    graph.addEdge(0, 1);
    graph.addEdge(0, 2);
    graph.addEdge(1, 3);
    graph.addEdge(2, 4);
    graph.bfs(0);
    // Output: bfs: 0 1 2 3 4
    graph.dfs(0);
    // Output: dfs: 0 1 3 2 4

    // Cross-language note:
    //   C++ unique_ptr owns each node — no GC; reverse rewires pointers in place.
    //   Python/TS/Java use object references; Rust uses Box<Option<Node>> for trees.
    //   Graph BFS uses std::queue here; DFS uses recursion (stack implicit in call frame).
    return 0;
}

File: rust/22_custom_data_structures.rs
// ============================================================
// Topic:  Custom Data Structures  (FOCUS: linked list, tree, graph)
// Summary: Box-backed list and BST; adjacency-list graph with BFS/DFS.
// Run:     rustc --edition 2021 rust/22_custom_data_structures.rs -o rust/22_custom_data_structures.exe && rust/22_custom_data_structures.exe
//
// KEY DIFFERENCES:
//   Python:  class Node with references; list built-in; no Box
//   TS/JS:   object references; no explicit ownership
//   Rust:    Box<T> for heap-owned nodes; Option<Box<T>> for nullable links
//   Java:    class Node { Node next; } — GC manages heap
//   C++:     raw pointers or unique_ptr for tree/list nodes
// ============================================================

use std::collections::VecDeque;

// ----- 1. Singly linked list (Box-backed nodes) -----
struct ListNode {
    val: i32,
    next: Option<Box<ListNode>>,
}

struct LinkedList {
    head: Option<Box<ListNode>>,
}

impl LinkedList {
    fn new() -> Self {
        LinkedList { head: None }
    }

    fn push_back(&mut self, val: i32) {
        let new_node = Box::new(ListNode { val, next: None });
        if self.head.is_none() {
            self.head = Some(new_node);
            return;
        }
        let mut cur = self.head.as_mut().unwrap();
        while cur.next.is_some() {
            cur = cur.next.as_mut().unwrap();
        }
        cur.next = Some(new_node);
    }

    fn to_vec(&self) -> Vec<i32> {
        let mut out = Vec::new();
        let mut cur = self.head.as_ref();
        while let Some(node) = cur {
            out.push(node.val);
            cur = node.next.as_ref();
        }
        out
    }

    fn reverse(&mut self) {
        let mut prev: Option<Box<ListNode>> = None;
        let mut cur = self.head.take();
        while let Some(mut node) = cur {
            let next = node.next.take();
            node.next = prev;
            prev = Some(node);
            cur = next;
        }
        self.head = prev;
    }
}

// ----- 2. Binary tree (Box-backed nodes) -----
struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

struct BinaryTree {
    root: Option<Box<TreeNode>>,
}

impl BinaryTree {
    // Manual build: root 1, left 2 (children 4, 5), right 3
    fn sample() -> Self {
        BinaryTree {
            root: Some(Box::new(TreeNode {
                val: 1,
                left: Some(Box::new(TreeNode {
                    val: 2,
                    left: Some(Box::new(TreeNode {
                        val: 4,
                        left: None,
                        right: None,
                    })),
                    right: Some(Box::new(TreeNode {
                        val: 5,
                        left: None,
                        right: None,
                    })),
                })),
                right: Some(Box::new(TreeNode {
                    val: 3,
                    left: None,
                    right: None,
                })),
            })),
        }
    }

    fn inorder(&self) -> Vec<i32> {
        let mut out = Vec::new();
        Self::inorder_node(self.root.as_ref(), &mut out);
        out
    }

    fn inorder_node(node: Option<&Box<TreeNode>>, out: &mut Vec<i32>) {
        if let Some(n) = node {
            Self::inorder_node(n.left.as_ref(), out);
            out.push(n.val);
            Self::inorder_node(n.right.as_ref(), out);
        }
    }

    fn preorder(&self) -> Vec<i32> {
        let mut out = Vec::new();
        Self::preorder_node(self.root.as_ref(), &mut out);
        out
    }

    fn preorder_node(node: Option<&Box<TreeNode>>, out: &mut Vec<i32>) {
        if let Some(n) = node {
            out.push(n.val);
            Self::preorder_node(n.left.as_ref(), out);
            Self::preorder_node(n.right.as_ref(), out);
        }
    }

    fn postorder(&self) -> Vec<i32> {
        let mut out = Vec::new();
        Self::postorder_node(self.root.as_ref(), &mut out);
        out
    }

    fn postorder_node(node: Option<&Box<TreeNode>>, out: &mut Vec<i32>) {
        if let Some(n) = node {
            Self::postorder_node(n.left.as_ref(), out);
            Self::postorder_node(n.right.as_ref(), out);
            out.push(n.val);
        }
    }
}

// ----- 3. Graph (adjacency list: Vec<Vec<usize>>) -----
struct Graph {
    adj: Vec<Vec<usize>>,
}

impl Graph {
    fn new(n: usize) -> Self {
        Graph {
            adj: vec![Vec::new(); n],
        }
    }

    fn add_edge(&mut self, u: usize, v: usize) {
        self.adj[u].push(v);
        self.adj[v].push(u);
    }

    fn bfs(&self, start: usize) -> Vec<usize> {
        let mut visited = vec![false; self.adj.len()];
        let mut queue = VecDeque::new();
        let mut out = Vec::new();

        visited[start] = true;
        queue.push_back(start);

        while let Some(node) = queue.pop_front() {
            out.push(node);
            for &next in &self.adj[node] {
                if !visited[next] {
                    visited[next] = true;
                    queue.push_back(next);
                }
            }
        }
        out
    }

    fn dfs(&self, start: usize) -> Vec<usize> {
        let mut visited = vec![false; self.adj.len()];
        let mut out = Vec::new();
        self.dfs_visit(start, &mut visited, &mut out);
        out
    }

    fn dfs_visit(&self, node: usize, visited: &mut [bool], out: &mut Vec<usize>) {
        visited[node] = true;
        out.push(node);
        for &next in &self.adj[node] {
            if !visited[next] {
                self.dfs_visit(next, visited, out);
            }
        }
    }
}

fn print_nums(label: &str, nums: &[i32]) {
    print!("{}:", label);
    for n in nums {
        print!(" {}", n);
    }
    println!();
}

fn print_usize(label: &str, nums: &[usize]) {
    print!("{}:", label);
    for n in nums {
        print!(" {}", n);
    }
    println!();
}

fn main() {
    // Linked list: 1 -> 2 -> 3 -> 4 -> 5
    let mut list = LinkedList::new();
    for v in [1, 2, 3, 4, 5] {
        list.push_back(v);
    }
    print_nums("list", &list.to_vec());
    list.reverse();
    print_nums("reversed", &list.to_vec());

    let tree = BinaryTree::sample();
    print_nums("inorder", &tree.inorder());
    print_nums("preorder", &tree.preorder());
    print_nums("postorder", &tree.postorder());

    // Undirected graph on 5 nodes (0..4)
    let mut graph = Graph::new(5);
    graph.add_edge(0, 1);
    graph.add_edge(0, 2);
    graph.add_edge(1, 3);
    graph.add_edge(1, 2);
    graph.add_edge(2, 4);

    print_usize("bfs", &graph.bfs(0));
    print_usize("dfs", &graph.dfs(0));
}